### ISSUE TITLE
signing error

### ISSUE THREAD

**Issue by Developer563** [Murmele/Gittyup#221]
I have configured the pgp keys
but I don't know why signing doesn't work.

**Comment by Murmele** [comment]
Hi @Developer563 can you explain more?

**Comment by exactly-one-kas** [comment]
Signing commits currently isn't implemented in Gittyup, I'm changing this to a feature request

(`libgit2` supports it via `git_commit_create_with_signature`, but we'll have to implement creating a signature ourself, probably using `gpgme`)

**Comment by Akselmo** [comment]
Hi, I'm interested in this feature and maybe could try my hand at implementing it. 

However, could the feature at first be just the signing process itself? Users could create the keys/signatures with other tools.

**Comment by exactly-one-kas** [comment]
Yes, just the signing would work
I'd say we simply require GPG to be installed and spawn `gpg --detach-sign --armor --local-user <committer email>` to create the signature

**Comment by Akselmo** [comment]
I was thinking a following plan:
- [ ] If signing is enabled in repo, use gpg to sign it
- [ ] Set gpg path in settings

But should the user be told somewhere that signing is enabled in the repository?

**Comment by exactly-one-kas** [comment]
Sounds good
Is `commit.gpgsign` the option to sign stuff? If so, putting that as a checkbox into the repository settings seems like a good idea
Also, using the commiter's email should probably be a fallback if `user.signingkey` is unset

**Comment by Akselmo** [comment]
Hey just an update, im still interested working on the feature, i just havent had the time yet due to getting a new job and life is busy.

I will get on this ASAP, but if someone else has more time than I do, im not stopping you :)

**Comment by exactly-one-kas** [comment]
Don't worry, take your time
Your life should take priority over volunteer work

**Comment by Akselmo** [comment]
`git_commit_create_with_signature` asks for `const char *commit_content` but the documentation doesn't mention what exactly is the content.

Is it `git_commit_body`?

**Comment by bonswouar** [comment]
Would love to see this feature implemented!

@Akselmo since it's been 13 days your might have your answer already
But just in case, according to the tests in libgit2 this `commit_content` is basically the whole commit content, like :
```
tree 4b825dc642cb6eb9a060e54bf8d69288fbee4904\n\
parent 8496071c1b46c854b31185ea97743be6a8774479\n\
author Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
committer Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
magicsig magic word: pretty please\n\
\n\
a simple commit which works\n";
```
See https://github.com/libgit2/libgit2/blob/ac0f2245510f6c75db1b1e7af7ca01c15dec26bc/tests/libgit2/commit/write.c#L349

**Comment by Akselmo** [comment]
> Would love to see this feature implemented!
> 
> @Akselmo since it's been 13 days your might have your answer already But just in case, according to the tests in libgit2 this `commit_content` is basically the whole commit content, like :
> 
> ```
> tree 4b825dc642cb6eb9a060e54bf8d69288fbee4904\n\
> parent 8496071c1b46c854b31185ea97743be6a8774479\n\
> author Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
> committer Ben Burkert <ben@benburkert.com> 1358451456 -0800\n\
> magicsig magic word: pretty please\n\
> \n\
> a simple commit which works\n";
> ```
> 
> See https://github.com/libgit2/libgit2/blob/ac0f2245510f6c75db1b1e7af7ca01c15dec26bc/tests/libgit2/commit/write.c#L349

Thanks! I have not been working on this much due to being busy with Life:tm:, but I'll see if I can get some work done on this this weekend :)

**Comment by Akselmo** [comment]
Hm, I don't know where to place the `git_commit_create_with_signature`  just to test it works at all.

I've tried to place it in `Repository.cpp` around line 629, but I'm not sure if this is the right place.

Can anyone give me a proper walkthrough what happens when user presses the commit button? I can't get debugger working with this either, GDB just crashes.

**Comment by bonswouar** [comment]
> I've tried to place it in Repository.cpp around line 629, but I'm not sure if this is the right place. Basically I'm lost in the codebase

@Akselmo I didn't dig much into the code but it seems you're on the right track, not sure if you've got access to everything for reconstructing the raw commit data from here though, you might need to create a new method with other parameters.

Unfortunately I haven't done any cpp for years and I don't know this project at all so I won't be able to help you much more :/

**Comment by exactly-one-kas** [comment]
@Akselmo `Repository::commit()` would be the right place, you can create another parameter that hands in the signature or the private key

**Comment by OdinVex** [comment]
For now, you can Edit Config and add:
[commit]
	gpgSign = true
[push]
	gpgSign = true
[tag]
	gpgSign = true

**Comment by Akselmo** [comment]
Once again I apologize for my slowness. :sweat_smile: Just wanted to say I haven't forgotten this.

Also, I am not against it if someone else can implement the solution before me.

**Comment by exactly-one-kas** [comment]
Don't worry - you're doing this for free and are free to take whatever time you want

**Comment by OdinVex** [comment]
I think I should clarify my post above, this works with an installed Gittyup (non-Flatpak, I have not tested Flatpak versions) as it uses my global git that has my GPG signing key in it and all.

**Comment by Akselmo** [comment]
Hey, I have unassigned myself, since I don't really use Gittyup anymore and I wanted to make it clear that anyone who has interest and more time to work on this can go for it.

I may give this another try, but I hope my questions may help anyone else who will attempt this.

Sorry I couldn't make this contribution happen (for now at least!). Gittyup is great software and I hope to see it improve in future. :)

**Comment by Murmele** [comment]
Hi, is it available in newer libgit2 versions? Currently I am working on updating the libgit2 version to make it upstream compatible again.
https://github.com/Murmele/Gittyup/pull/153
Maybe you can check there

**Comment by NotCoolKiwi** [comment]
I would also be interested in seeing this implemented. I like this project overall, but my org requires us to sign our commits, so unfortunately I can't use it until then.

**Comment by rimas-kudelis** [comment]
This was mentioned above in passing, but somehow only once.

It seems that [GPGME library](https://www.gnupg.org/software/gpgme/index.html) is the suggested way of using GPG from within other applications, not executing the `gpg` binary.

**Comment by RokeJulianLockhart** [comment]
Is this a duplicate of https://github.com/Murmele/Gittyup/issues/121#issue-1185637221?
